### PR TITLE
Automate Failover across regions

### PR DESCRIPTION
**Pull request by davidrusu**
This PR implements zookeeper based failover, our approach uses the Zookeeper barrier pattern: https://zookeeper.apache.org/doc/r3.1.2/recipes.html#sc_recipes_eventHandles

### The Plan:

As the primary db2kafka instance:
- Before starting work, establish a zookeeper session and create an ephemeral node ie. `/failover_us-west-2`
- we then start doing work

As the secondary db2kafka instance:
- Before starting work, establish a zookeeper session and check for the existence of the primary's ephemeral node.
- if the node exists, it will wait until it receives a notification that the node was deleted.
- if the node doesn't exist, it will start doing work
- when we are not doing work and we receive a notification that the primary's node was deleted, we start doing work.
- when we are doing work and we receive a notification that the primary's node was created, 
 the we will crash (and hopefully be reinitialized by its process monitor).

### CODE REVIEW

File: lib/failover/failover.ex [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+  :failover_stop
+    This event will fire if this instance is doing work and one of these happen:
+    - Our zk session has expired
+    - We've lost our zk connection.
+    - You are a failover instance and we've detected the primary instance is up
+
+  In any of these cases, your service should stop doing work, else you risk
+  having both primary and failover instances doing work at the same time.
+  """
```
Comment (DWvanGeest) [inline lib/failover/failover.ex]: This module has two different state machines, one for the primary and one for the failover. This makes the logic a bit hard to follow and errors harder to spot. For example `def handle_info({:node_created, path}, state)` is only ever called when this instance is a failover I believe, but there's nothing in here that tells me that without having to trace out the state machine. 

Splitting this module into Primary and Failover pieces would separate the two state machines and make it much easier to follow. The common ZK stuff and `start_app` etc could go in a third module required/imported/whatever-it's-called by the others.

What do you think? Do-able?

File: lib/failover/failover.ex [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+  def handle_info({:node_deleted, path}, state) do
+    Logger.info("Barrier came down: #{inspect path}, starting failover region")
+    prepare_failover(state)
+  end
+
+  defp post_zk_connect_init(state) do
+    %{primary_region: primary_region, region: region} = state
+
+    if primary_region == region do
```
Comment (DWvanGeest) [inline lib/failover/failover.ex]: Basically this comparison would go into the parent of this process, and the parent would start `Primary` or `Failover` based on it.

File: test/failover/failover_test.exs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    
+    assert GenServer.call(state[:instance_pid], :is_working) == true
+
+    {:noreply, state} = with_mock :erlzk_conn, [mock_zk_exists(:exists)] do
+      Failover.handle_info({:node_created, '/failover'}, state)
+    end
+    
+    assert GenServer.call(state[:instance_pid], :is_working) == false
+  end
+end
```
Comment (DWvanGeest) [inline test/failover/failover_test.exs]: Nice tests!

File: test/db2kafka/topic_health_tracker_test.exs [hunk 4]
```diff
@@ -41,46 +41,46 @@ defmodule Db2Kafka.TopicHealthTracker.Test do
...
 
     assert elem(response, 0) == :noreply
     state = elem(response, 1)
     assert state.topic == topic
     assert state.partitions_status[1] == part_1_time
-    assert abs(state.partitions_status[0] - current_unix_time) < 30
+    assert abs(state.partitions_status[0] - current_unix_time()) < 30
 
-    assert_not_terminated
+    assert_not_terminated()
```
Comment (DWvanGeest) [inline test/db2kafka/topic_health_tracker_test.exs]: These have been annoying me too, thanks for fixing them.

File: lib/failover/failover.ex [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        # on restart to reinitialize us so we don't need these features
+        disable_watch_auto_reset: true,
+        disable_expire_reconnect: true,
+      ]
+
+    zk_host =  Application.get_env(:db2kafka, :zk_host) |> to_charlist
+    zk_port = Application.get_env(:db2kafka, :zk_port) |> String.to_integer
+    zk_hosts = [{zk_host, zk_port}]
+    zk_session_timeout = 10000
+    {:ok, zk} = :erlzk_conn.start_link(zk_hosts, zk_session_timeout, options)
```
Comment (davidrusu) [inline lib/failover/failover.ex]: No, this won't act as a seed node. I'll make the change to allow passing in multiple zk hosts in the config.

File: lib/failover/failover.ex [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        Logger.info("Created barrier #{path}, starting primary region")
+        {:ok, state} = start_app(state)
+        {:noreply, state}
+      {:error, reason} ->
+        # For all the error values see:
+        # https://github.com/huaban/erlzk/blob/master/src/erlzk.erl#L154
+        #
+        # The error we will most likely receive is :node_exists.
+        # This can happen if the instance dies and restarts before the zk
+        # session expires. In that case the barrier node will still exist
```
Comment (davidrusu) [inline lib/failover/failover.ex]: The session timeout we are using is 10 seconds.

File: lib/failover/failover.ex [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+  defp prepare_failover(%{zk: zk}=state) do
+    Logger.info("Checking state of barrier #{inspect barrier_path(state)}")
+    case watch_barrier(zk, barrier_path(state)) do
+      {:ok, _stat} ->
+        Logger.info("Barrier exists, waiting for barrier to go down")
+        {:noreply, state}
+      {:error, :no_node} ->
+        Logger.info("No barrier, this implies primary is down")
+        {:ok, state} = start_app(state)
```
Comment (davidrusu) [inline lib/failover/failover.ex]: Yep, the watch is active here. I'm not a fan of this part of `erlzk`'s api. `:no_node` is a valid response, it shouldn't be treated as an `:error`.

File: lib/failover/failover.ex (hunk 5 above)
Comment (DWvanGeest) [inline lib/failover/failover.ex]: Ah. So if that one ZK node goes down, ZK connection will be lost even though the app instance can still do work and could still maintain the barrier via other nodes (?).

File: lib/failover/failover.ex (hunk 6 above)
Comment (DWvanGeest) [inline lib/failover/failover.ex]: :+1:

File: lib/failover/failover.ex (hunk 5 above)
Comment (davidrusu) [inline lib/failover/failover.ex]: Yup exactly, I have a change coming to fix this

File: lib/db2kafka/partition_processor.ex [hunk 8]
```diff
@@ -3,7 +3,6 @@ defmodule Db2Kafka.PartitionProcessor do
   use Bitwise
   require Logger
 
-  @full_trip_metric "db2kafka.full_trip"
```
Comment (davidrusu) [inline lib/db2kafka/partition_processor.ex]: not used

File: lib/db2kafka/record_buffer.ex [hunk 9]
```diff
@@ -7,7 +7,6 @@ defmodule Db2Kafka.RecordBuffer do
   @max_batch_size 3_000
 
   @records_served_metric "db2kafka.buffer.records_served"
-  @duplicates_metric "db2kafka.buffer.duplicates"
```
Comment (davidrusu) [inline lib/db2kafka/record_buffer.ex]: not used

File: lib/failover/failover.ex (hunk 1 above)
Comment (davidrusu) [inline lib/failover/failover.ex]: Good point, spent a bit of time to refactor this into smaller modules

File: test/test_helper.exs [hunk 10]
```diff
@@ -65,8 +65,6 @@ defmodule RecordCreator do
   use GenServer
   require Logger
 
-  @table "outbound_kafka_queue"
-
```
Comment (davidrusu) [inline test/test_helper.exs]: not used

File: lib/failover/zk_helper.ex [hunk 11]
```diff
@@ -0,0 +1,85 @@
+defmodule Failover.ZKHelper do
+  @moduledoc """
+  This module provides a GenServer that attempts to simplify common interactions with the erlzk client in the context of failover.
+
+  In particular, it provides the callbacks to notify when we have the gurantee of an active ZK session:
+
+  handle_cast(:failover_safe_to_start, state)      # sent once our ZK session is active
+  handle_cast(:failover_unsafe_to_continue, state) # sent when our ZK session has expired or our ZK connection has died
```
Comment (DWvanGeest) [inline lib/failover/zk_helper.ex]: I definitely like the idea of this module - isolate the library choice and abstract the barrier pattern ZK stuff. However I'm wondering if it needs to know (or actually does know) anything about failover, which has logic encapsulated elsewhere. Basically I think this module should know about building barriers in ZK only.

All that would mean is changing the callback naming somewhat. `:failover_safe_to_start` could just be `:connected` or `:session_ready` or something, and `:failover_unsafe_to_continue` would be `:disconnected` or `:session_lost`. Then it's the responsibility of the Failover modules to react to those events accordingly. Theoretically you could use this module for any kind of barrier application then.

File: lib/failover/secondary.ex [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    {:ok, state} = Failover.stop_app(state)
+    {:ok, state} = prepare_failover(state)
+    {:noreply, state}
+  end
+
+  def handle_info({:node_deleted, path}, state) do
+    Logger.info("Barrier came down: #{inspect path}, starting failover region")
+    {:ok, state} = prepare_failover(state)
+    {:noreply, state}
+  end
```
Comment (DWvanGeest) [inline lib/failover/secondary.ex]: These two handlers are a bit of a [leaky abstraction](https://en.wikipedia.org/wiki/Leaky_abstraction), in that this module now knows about the callbacks the underlying ZK lib exposes, even though `ZkHelper` really should encapsulate that.

How about `ZkHelper` has callbacks like `:barrier_down` and `:barrier_up`? Internally they would be just be a translation from `:node_deleted` etc.

File: lib/failover/zk_helper.ex [hunk 13]
```diff
@@ -0,0 +1,85 @@
+defmodule Failover.ZKHelper do
```
Comment (DWvanGeest) [inline lib/failover/zk_helper.ex]: The more I think about this, the more I think this module is specifically a `ZkBarrierManager` or something of that sort, in that it's specific to ZK and the Barrier pattern.

File: lib/failover/secondary.ex [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  def handle_info({:node_created, path}, state) do
+    Logger.info("Barrier came up: #{inspect path}, stopping failover region")
+    {:ok, state} = Failover.stop_app(state)
+    {:ok, state} = prepare_failover(state)
+    {:noreply, state}
+  end
+
+  def handle_info({:node_deleted, path}, state) do
+    Logger.info("Barrier came down: #{inspect path}, starting failover region")
+    {:ok, state} = prepare_failover(state)
```
Comment (DWvanGeest) [inline lib/failover/secondary.ex]: Are you no longer watching the barrier once this event is received? Calling `prepare_failover` does the watch again, so if the watch ends on this event it's all good. But, if the watch doesn't end it seems like you have two watches going (even eventually more).

File: config/config.exs [hunk 15]
```diff
@@ -10,7 +10,17 @@ config :db2kafka,
   db_username:  System.get_env("DB_USER") || "root",                     # MySQL username
   db_password:  System.get_env("DB_PASS") || "",                         # MySQL password
   db_name:      System.get_env("DB_NAME") || "db2kafka",                 # MySQL database
-  topics:      (System.get_env("TOPICS") || "foo") |> String.split(",")  # Comma-separated list of topics
+  topics:      (System.get_env("TOPICS") || "foo") |> String.split(","), # Comma-separated list of topics
+
+  # Failover configuration
+  zk_hosts: [                                                            # List known ZK hosts
+    { System.get_env("ZK_HOST") || "localhost",
```
Comment (DWvanGeest) [inline config/config.exs]: It's a bit unclear to me what's happening here. How do you specify multiple hosts? It appears (based on naming and structure) that you can specify a single host in `ZK_HOST`, at which point I don't understand the point of making `zk_hosts` an array.

Maybe use an env variable like `ZK_HOSTS` which is in the format `host1:port1,host2:port2`, and then just transform it into the expected variables for the driver lib in `ZkHelper`?

File: lib/failover/secondary.ex (hunk 14 above)
Comment (davidrusu) [inline lib/failover/secondary.ex]: Zookeeper watches fire at most once, if we receive the `:node_deleted` message this means the watch has fired and we need to set it again before continuing.

See: https://zookeeper.apache.org/doc/trunk/zookeeperProgrammers.html#ch_zkWatches

Comment (DWvanGeest) [inline lib/failover/secondary.ex]: :+1: TIL

File: config/config.exs (hunk 15 above)
Comment (davidrusu) [inline config/config.exs]: We won't be using env variables in production, these env variables are to ease development only. We do the same thing below with Kafka brokers: https://github.com/PagerDuty/db2kafka/pull/8/files/62f0602e778b7b0b3f65c2ea1156e207e297ff21#diff-de9210b77e43a621a256d7a37fc5f50aR26

I say, no need to make this complicated if it won't be used.

File: lib/failover/zk_helper.ex (hunk 11 above)
Comment (davidrusu) [inline lib/failover/zk_helper.ex]: I added the `failover` prefix to these messages to act as a sort of namespace so that other modules don't get confused (also makes grepping easier as we fewer false positives).

Yeah... I'll see about moving more of the barrier logic into this module. I was hesitant to do to much in here as I wanted to keep it as stateless as possible, but I think there may be a way to move all of the barrier logic here and keep this module simple.

File: lib/failover/secondary.ex (hunk 12 above)
Comment (davidrusu) [inline lib/failover/secondary.ex]: I'll look into it, good call

File: lib/failover/zk_helper.ex (hunk 11 above)
Comment (DWvanGeest) [inline lib/failover/zk_helper.ex]: To be clear, I don't think you need to move any logic. Just rename the callbacks so they aren't `failover`-specific.

File: config/config.exs (hunk 15 above)
Comment (DWvanGeest) [inline config/config.exs]: OK. I guess I was a bit confused about how the library user would use this, but we're assuming that this variable is overridden entirely by the application. Might be worth a README entry, but not blocking this PR.
